src/mpid/ch4/netmod/ofi/ofi_am_impl.c

```diff
@@ -11,71 +11,20 @@
-    bool do_eager = false;
 
     MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDI_OFI_DEFERRED_AM_ISEND_ISSUE);
     MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDI_OFI_DEFERRED_AM_ISEND_ISSUE);
 
-    MPIDI_Datatype_get_info(dreq->count, dreq->datatype, dt_contig, data_sz, dt_ptr, dt_true_lb);
-    need_packing = dt_contig ? false : true;
-    send_buf = (char *) dreq->buf + dt_true_lb;
-
-    do_eager =
-        (MPIDI_OFI_AMREQUEST_HDR(dreq->sreq, am_hdr_sz) + dreq->data_sz
-         + sizeof(MPIDI_OFI_am_header_t) <= MPIDI_OFI_DEFAULT_SHORT_SEND_SIZE);
-
-    MPL_pointer_attr_t attr;
-    MPIR_GPU_query_pointer_attr(dreq->buf, &attr);
-    if (attr.type == MPL_GPU_POINTER_DEV && !MPIDI_OFI_ENABLE_HMEM) {
-        /* Force packing of GPU buffer in host memory */
-        need_packing = true;
-    }
-
-    if (need_packing) {
-        /* FIXME: currently we always do packing, also for high density types. However,
-         * we should not do packing unless needed. Also, for large low-density types
-         * we should not allocate the entire buffer and do the packing at once. */
-        /* TODO: (1) Skip packing for high-density datatypes;
-         *       (2) Pipeline allocation for low-density datatypes; */
-        if (do_eager) {
-            mpi_errno = MPIDU_genq_private_pool_alloc_cell(MPIDI_OFI_global.pack_buf_pool,
-                                                           (void **) &send_buf);
-            if (send_buf == NULL) {
-                /* no buffer available, suppress error and exit */
-                mpi_errno = MPI_SUCCESS;
-                goto fn_exit;
-            }
-            MPIR_ERR_CHECK(mpi_errno);
-        } else {
-            /* only for large message RDMA read */
-            MPL_gpu_malloc_host((void **) &send_buf, data_sz);
-        }
-        mpi_errno = MPIR_Typerep_pack(dreq->buf, dreq->count, dreq->datatype, 0, send_buf,
-                                      dreq->data_sz, &last);
-        MPIR_ERR_CHECK(mpi_errno);
-
-        MPIDI_OFI_AMREQUEST_HDR(dreq->sreq, pack_buffer) = send_buf;
-    } else {
-        MPIDI_OFI_AMREQUEST_HDR(dreq->sreq, pack_buffer) = NULL;
-    }
-
-    if (do_eager) {
-        mpi_errno = MPIDI_OFI_am_isend_short(dreq->rank, dreq->comm, dreq->handler_id, send_buf,
-                                             dreq->data_sz, dreq->sreq);
-    } else {
-        mpi_errno = MPIDI_OFI_am_isend_long(dreq->rank, dreq->comm, dreq->handler_id, send_buf,
-                                            dreq->data_sz, dreq->sreq);
+    switch (dreq->op) {
+        case MPIDI_OFI_DEFERRED_AM_OP__ISEND_EAGER:
+            mpi_errno = MPIDI_OFI_do_am_isend_eager(dreq->rank, dreq->comm, dreq->handler_id, NULL,
+                                                    0, dreq->buf, dreq->count, dreq->datatype,
+                                                    dreq->sreq, true);
+            break;
+        case MPIDI_OFI_DEFERRED_AM_OP__ISEND_PIPELINE:
+        default:
+            MPIR_Assert(0);
```

Comment:
Now there is just this single function in this file and it is rather short, how about we move it inside `handle_deferred_ops` in `ofi_progress.c`?